### PR TITLE
[MI-3235] Add config validations

### PR DESCRIPTION
**Pull request by ayusht2810**

### CODE REVIEW

File: constants/constants.go [hunk 1]
```diff
@@ -12,10 +12,26 @@ const (
...
+	ErrorEmptyAdminEmail         = "admin email should not be empty"
+	ErrorEmptyAdminPassword      = "admin password should not be empty"
+	ErrorEmptyUsername           = "username should not be empty"
+	ErrorEmptyUserPassword       = "user password should not be empty"
+	ErrorEmptyUserEmailL         = "user email should not be empty"
+	ErrorEmptyChannelDisplayName = "channel display name should not be empty"
+	ErrorEmptyChannelName        = "channel name should not be empty"
+	ErrorEmptyChannelType        = "channel type should not be empty"
+	ErrorEmptyMMTeamName         = "mattermost team name should not be empty"
+	ErrorEmptyMSTeamsTeamID      = "ms teams team ID not be empty"
```
Comment (raghavaggarwal2308) [inline constants/constants.go]:
```suggestion
	ErrorEmptyMSTeamsTeamID      = "ms teams team ID should not be empty"
```

File: constants/constants.go [hunk 2]
```diff
@@ -12,10 +12,26 @@ const (
...
 )
+
+// Validation errors
+const (
+	ErrorEmptyServerURL          = "server URL should not be empty"
+	ErrorEmptyAdminEmail         = "admin email should not be empty"
+	ErrorEmptyAdminPassword      = "admin password should not be empty"
+	ErrorEmptyUsername           = "username should not be empty"
+	ErrorEmptyUserPassword       = "user password should not be empty"
+	ErrorEmptyUserEmailL         = "user email should not be empty"
```
Comment (raghavaggarwal2308) [inline constants/constants.go]:
```suggestion
	ErrorEmptyUserEmail         = "user email should not be empty"
```

File: k6/createPosts.js [hunk 3]
```diff
@@ -8,6 +8,18 @@ export const options = {
     duration: config.PostsConfiguration.Duration,
 }
 
+export function setup() {
+	if (config.PostsConfiguration.MaxWordsCount <= 0) {
+        console.error("Error in validating the posts configuration:", "max words count should not be empty and greater than 0");
```
Comment (raghavaggarwal2308) [inline k6/createPosts.js]:
```suggestion
        console.error("Error in validating the posts configuration:", "max words count should be greater than 0");
```

Comment (raghavaggarwal2308) [inline k6/createPosts.js]: indentation

File: k6/createPosts.js [hunk 4]
```diff
@@ -8,6 +8,18 @@ export const options = {
...
+		return;
+	}
+
+	if (config.PostsConfiguration.MaxWordLength <= 0) {
+        console.error("Error in validating the posts configuration:", "max words length not be empty and greater than 0");
```
Comment (raghavaggarwal2308) [inline k6/createPosts.js]:
```suggestion
        console.error("Error in validating the posts configuration:", "max word length should be greater than 0");
```

File: main.go [hunk 5]
```diff
@@ -24,14 +23,29 @@ func main() {
 		return
 	}
```
Comment (raghavaggarwal2308) [inline main.go:25]:
```suggestion
```

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: k6/createPosts.js (hunk 3 above)
Comment (ayusht2810) [reply k6/createPosts.js]: it is correct in the code

**Review by raghavaggarwal2308 — APPROVED** [review]

File: k6/createPosts.js [hunk 6]
```diff
@@ -8,6 +8,18 @@ export const options = {
     duration: config.PostsConfiguration.Duration,
 }
 
+export function setup() {
+	if (config.PostsConfiguration.MaxWordsCount <= 0) {
+        console.error("Error in validating the posts configuration:", "max words count should be greater than 0");
```
Comment (Kshitij-Katiyar) [inline k6/createPosts.js]:
```suggestion
        console.error("Error in validating the posts configuration:", "max word count should be greater than 0");
```

File: main.go [hunk 7]
```diff
@@ -43,7 +57,5 @@ func main() {
 		logger.Error("failed to run the script", zap.String("arg", args[1]), zap.Error(err))
 	}
 
-	if err := logger.Sync(); err != nil {
-		panic(err)
-	}
+	_ = logger.Sync()
```
Comment (Kshitij-Katiyar) [inline main.go:60]: why this change ?

**Review by Kshitij-Katiyar — APPROVED** [review]

File: main.go (hunk 7 above)
Comment (ayusht2810) [reply main.go:60]: this was removed as `logger.Sync()` returns an error whenever we log an error from Zap, which causes panic.
You can read more about it here
https://github.com/uber-go/zap/issues/328#fromHistory
https://github.com/uber-go/zap/pull/347

File: constants/constants.go [hunk 8]
```diff
@@ -12,10 +12,26 @@ const (
...
+// Validation errors
+const (
+	ErrorEmptyServerURL          = "server URL should not be empty"
+	ErrorEmptyAdminEmail         = "admin email should not be empty"
+	ErrorEmptyAdminPassword      = "admin password should not be empty"
+	ErrorEmptyUsername           = "username should not be empty"
+	ErrorEmptyUserPassword       = "user password should not be empty"
+	ErrorEmptyUserEmail          = "user email should not be empty"
+	ErrorEmptyChannelDisplayName = "channel display name should not be empty"
+	ErrorEmptyChannelName        = "channel name should not be empty"
```
Comment (manojmalik20) [inline constants/constants.go]: I think we should rename this config variable to `channel slug name` instead of `channel name`.

File: serializers/config.go [hunk 9]
```diff
@@ -29,5 +37,80 @@ type ChannelsConfiguration struct {
...
+
+	c.ConnectionConfiguration.ServerURL = strings.TrimRight(strings.TrimSpace(c.ConnectionConfiguration.ServerURL), "/")
+
+	return nil
+}
+
+func (c *Config) IsUsersConfigurationValid() error {
+	for _, user := range c.UsersConfiguration {
+		if user.Username == "" {
+			return errors.New(constants.ErrorEmptyUsername)
```
Comment (manojmalik20) [inline serializers/config.go]: We should also mention in the error that at what index we are talking about empty username.

File: serializers/config.go [hunk 10]
```diff
@@ -29,5 +37,80 @@ type ChannelsConfiguration struct {
...
+
+		if user.Email == "" {
+			return errors.New(constants.ErrorEmptyUserEmail)
+		}
+
+		if user.Password == "" {
+			return errors.New(constants.ErrorEmptyUserPassword)
+		}
+
+		user.Email = strings.TrimSpace(user.Email)
```
Comment (manojmalik20) [inline serializers/config.go:82]: We should also do this for username, right? And for all such fields.

**Review by manojmalik20 — APPROVED** [review]
